homeassistant/components/knx/expose.py

```diff
@@ -116,11 +113,11 @@ def shutdown(self) -> None:
     def _get_expose_value(self, state: State | None) -> StateType:
         """Extract value from state."""
         if state is None or state.state in (STATE_UNKNOWN, STATE_UNAVAILABLE):
-            return None
+            return self.expose_default
         value = (
             state.state
             if self.expose_attribute is None
-            else state.attributes.get(self.expose_attribute)
+            else state.attributes.get(self.expose_attribute, self.expose_default)
```

Comment:
This is odd behavior imho.

If the default is set on line 116 (not exposing an attribute), it is just accepted directly, while if an attribute returns `None`, the default exposed value gets processed.

So should it be processed? If so, line 116 is wrong, if not, line 120 is wrong.

Reply:
Oh, right, thanks. It should always get processed.

Reply:
thanks for reviewing, but sorry, can't follow. If the sate is unknown in line 115, the attribute value also should be unknown, so directly returning the default value in 116 seems correct to me.

And the result of line 117-120 (state value or attribute value) is processed down the line. Or what do I miss here?

Reply:
If the attribute is None, it returns the default exposed default as a value, which is next taken down the method for further processing.

The difference is that in the first case on line 116, that further processing doesn't happen. So, does the default value need to go down the method or not? As those cases are now different, while both dealing with the same default value.

Reply:
Yes, the exposed value used in 
- `self.device.sensor_value.value` (`_init_expose_state()`) and 
- `await self.device.set(value)` (`async_set_knx_value()`)
should always be casted to the proper type (Line 122-132 - `_get_expose_value()`)

Reply:
oh right. Good catch. @farmio should I do the type sanitizing in `async_set_knx_value()` then, or keep it in `_get_expose_value()` and refactor that?

Reply:
would that work for you guys?
```python
        if state is None or state.state in (STATE_UNKNOWN, STATE_UNAVAILABLE):
            value = self.expose_default
        else:
            value = (
                state.state
                if self.expose_attribute is None
                else state.attributes.get(self.expose_attribute, self.expose_default)
            )
        if self.type == "binary":
            if value in (1, STATE_ON, "True"):
                return True
            if value in (0, STATE_OFF, "False"):
                return False
        if (
...
```

Reply:
However you prefer. Just make sure `_async_entity_changed()` can compare a default value with a non-default value.
Your suggestion should be fine.